chocolatey/choco
# (#2660) Add runtime check for compatibility with installed Chocolatey Licensed Extension

## Description Of Changes

In order to provide as much support to users of Chocolatey as possible,
we have added the concept of a minimum required Chocolatey version
attribute to the Chocolatey Licensed Assembly.  When reading this
assembly (if present), we will extract this attribute, and then use it
to make assertions about whether things are currently known to be
compatible.

When it is found that they are not compatible, a warning is shown
before and after the command execution to try to make it as clear as
possible that things might not work as expected.

A new feature flag has been introduced, disableCompatiblityChecks,
which a user can opt-in to, to prevent the warnings from showing, but
this is disabled by default.  The warning shows how this feature can be
turned on if they are sure that they know what they are doing.

In addition, a command line option has been added to allow by passing
these warnings without having to enable the feature, with the
skip-compatibility-checks option.

In addition, due to the fact that we can, and do, load the Chocolatey
Licensed Assembly from memory, the Location property can be null.  When
this is the case, and it hasn't been possible to extract the version
information, we reach into the custom attributes for the assembly to
pull out the information, as this is needed to assert on the
compatibility of the versions.

## Motivation and Context

This is required to protect people when there are known incompatibilities
with the installed version of the Chocolatey Licensed Extension.  This check
needs to happen before reflecting into the assembly to override the IoC
registrations, which is why this check needs to be done in Chocolatey CLI.

## Testing

1. Get a system that has Chocolatey 0.12.1 and Chocolatey Licensed Extension v3.2.0 installed
1. Run choco list -lo and ensure that it outputs as expected
1. Install the built alpha package from this PR
1. Run choco list -lo and see that there is a new warning before and after the command has executed
1. Run choco list -lo --skip-compatibility-checks and see that the warnings are not shown
1. Run choco feature enable --name="'disableCompatibilityChecks'"
1. Run choco list -lo and see that the warnings are not shown
1. Upgrade to the latest Chocolatey Licensed Extension
1. Run choco list -lo and see that the warnings are not shown
1. Run choco feature enable --name="'disableCompatibilityChecks'"
1. Run choco list -lo and see that the warnings are not shown

## Change Types Made


## Related Issue

Fixes #2660 

## Change Checklist
